### PR TITLE
sql: introduce MVCC-compliant index backfiller

### PR DESCRIPTION
**Pull request by stevendanna**
Previously, the index backfilling process depended upon non-MVCC
compliant AddSSTable calls which potentially rewrote previously read
historical values.

To support an MVCC-compliant AddSSTable that writes at the _current_
timestamp, this change implements a new backfilling process described
in the following RFC:

https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/20211004_incremental_index_backfiller.md

In summary, the new index backfilling process depends on backfilling
the new index when it is in a BACKFILLING state (added in #72281). In
this state it receives no writes or deletes. Writes that occur during
the backfilling process are captured by a "temporary index."  This
temporary index uses the DeletePreservingEncoding to ensure it
captures deletes as well as writes.

After the of bulk backfill using the MVCC-compliant AddSSTable, the
index is moved into a MERGING state
(added in #75663) in which it receives writes and deletes. Writes
previously captured by the temporary index are then transactionally
merged into the newly added index.

This feature is currently behind a new boolean cluster setting which
default to true. Schema changes that contains both old and new-style
backfills are rejected.

Reverting the default to false will require updating various tests
since many tests depend on the exact index IDs of newly added indexes.

Release note: None

### CODE REVIEW

**Comment by stevendanna** [comment]
@dt There are still some TODOs here, but I was wondering if you want to give this a review pass?

File: pkg/sql/schema_changer.go
Comment (dt) [inline pkg/sql/schema_changer.go:1193]: Is the all caps to style? fine with it if it is, just checking

**Review by dt — COMMENTED** [review]
I did a once through but defer to a schema reviewer on most of this that is pretty much pure schema code. My only real question is about if we actually care about initialReadTimestamp

File: pkg/sql/mvcc_backfiller.go
Comment (stevendanna) [reply pkg/sql/mvcc_backfiller.go:75]: I believe it is required for correctness. I've explained the case it is handling in the commit message, but can also put a comment here.

File: pkg/jobs/jobspb/jobs.proto
Comment (stevendanna) [reply pkg/jobs/jobspb/jobs.proto:653]: Yup, that is the only user of this.

File: pkg/sql/schema_changer.go
Comment (stevendanna) [reply pkg/sql/schema_changer.go:1193]: Thanks for catching this, I meant to fix it up, I made it all caps to make it easier to spot in assertion failures in some of the logic tests and then didn't change it back.

File: pkg/sql/schema_changer.go
Comment (stevendanna) [reply pkg/sql/schema_changer.go:2671]: No real reason, this was a recent addition to the schema changer so I should probably follow up with schema here.

File: pkg/sql/mvcc_backfiller.go
Comment (dt) [reply pkg/sql/mvcc_backfiller.go:75]: Ah. Okay, caught up on the commit message and now see why it's here. Let me ponder a bit, since I'm not sure this will work on top of restored data where we forward all timestamps.

Comment (stevendanna) [reply pkg/sql/mvcc_backfiller.go:75]: Yeah.  Another option (echoing from slack) is to persist the timestamp in the body of a preserved delete so that it could survive the restore.

File: pkg/jobs/jobspb/jobs.proto
Comment (stevendanna) [reply pkg/jobs/jobspb/jobs.proto]: This has been made unnecessary by not issuing delete-preserving deletes when the index is in DeleteOnly. See https://github.com/cockroachdb/cockroach/pull/75720

File: pkg/sql/mvcc_backfiller.go
Comment (stevendanna) [reply pkg/sql/mvcc_backfiller.go]: This has been made unnecessary by not issuing delete-preserving deletes when the index is in DeleteOnly. See #75720

File: pkg/sql/backfill.go
Comment (ajwerner) [inline pkg/sql/backfill.go:1859]: nit: copatible

File: pkg/sql/catalog/tabledesc/index.go
Comment (ajwerner) [inline pkg/sql/catalog/tabledesc/index.go:356]: ➕ it might be nice to be able to associate the two indexes.

File: pkg/sql/catalog/tabledesc/structured.go
Comment (ajwerner) [inline pkg/sql/catalog/tabledesc/structured.go:2150]: does this need version gating?

File: pkg/sql/execinfrapb/processors_bulk_io.proto
Comment (ajwerner) [inline pkg/sql/execinfrapb/processors_bulk_io.proto:78]: nit: indentation

File: pkg/sql/mvcc_backfiller.go
Comment (ajwerner) [inline pkg/sql/mvcc_backfiller.go:188]: let's file an issue for this?

File: pkg/sql/schema_changer.go
Comment (ajwerner) [inline pkg/sql/schema_changer.go:1107]: this and the stepping twice all feels a little brittle. is there a test which resumes between the two steps?

**Review by ajwerner — COMMENTED** [review]
This PR wasn't very surprising, which is good. On this first pass, I mostly just had nits and a couple of questions. I'm inclined to say we should merge this with some urgency and then keep iterating.

File: pkg/sql/schema_changer.go
Comment (stevendanna) [reply pkg/sql/schema_changer.go:1111]: I've added a test that randomly pauses before calls to `(*schemaChanger).txn` and then resume the job and makes sure it succeeds. It's a little slow so it only does one pause by default, but locally I've run it after every txn:

```
--- PASS: TestPauseBeforeRandomDescTxn (23.87s)                                                                               
    --- PASS: TestPauseBeforeRandomDescTxn/create_index_pause_at_txn_1 (1.77s)
    --- PASS: TestPauseBeforeRandomDescTxn/create_index_pause_at_txn_2 (1.78s)                                        
    --- PASS: TestPauseBeforeRandomDescTxn/create_index_pause_at_txn_3 (1.74s)                                                
    --- PASS: TestPauseBeforeRandomDescTxn/create_index_pause_at_txn_4 (1.80s)                                                
    --- PASS: TestPauseBeforeRandomDescTxn/create_index_pause_at_txn_5 (1.91s)                                                
    --- PASS: TestPauseBeforeRandomDescTxn/create_index_pause_at_txn_6 (1.72s)                                        
    --- PASS: TestPauseBeforeRandomDescTxn/create_index_pause_at_txn_7 (1.77s)                                                
    --- PASS: TestPauseBeforeRandomDescTxn/create_index_pause_at_txn_8 (1.72s)                                                
    --- PASS: TestPauseBeforeRandomDescTxn/create_index_pause_at_txn_9 (1.67s)                                                                                                                                                                              
    --- PASS: TestPauseBeforeRandomDescTxn/create_index_pause_at_txn_10 (1.82s)                                               
    --- PASS: TestPauseBeforeRandomDescTxn/create_index_pause_at_txn_11 (1.87s)                                               
    --- PASS: TestPauseBeforeRandomDescTxn/create_index_pause_at_txn_12 (1.94s)
    --- PASS: TestPauseBeforeRandomDescTxn/create_index_pause_at_txn_13 (1.74s)   
```

I think we need to add more assertions that we don't have any unexpected cruft and it would also be nice to add some concurrent operations, but hopefully this gets at the basic problem you are pointing out here.

File: pkg/sql/backfill/mvcc_index_merger.go
Comment (ajwerner) [inline pkg/sql/backfill/mvcc_index_merger.go:217]: If I were being nit-picky, I'd encourage you to use the lower-level APIs. These high-level APIs are sort of garbage. I'd consider something like:

```
var ba roachpb.BatchRequest
	ba.TargetBytes = 16 << 10
	ba.MaxSpanRequestKeys = chunkSize
	ba.Add(&roachpb.ScanRequest{
		RequestHeader: roachpb.RequestHeader{
			Key:    key,
			EndKey: endKey,
		},
		ScanFormat: roachpb.KEY_VALUES,
	})
	br, pErr := txn.Send(ctx, ba)
	if pErr != nil {
		return pErr.GoError()
	}
	resp := br.Responses[0].GetScan()
	for _, row := range resp.Rows {
		
	}
```
Otherwise you're liable to pull a lot more data than you wanted.

File: pkg/sql/distsql_physical_planner.go
Comment (ajwerner) [inline pkg/sql/distsql_physical_planner.go:987]: it feels like you may as well just check that userData and spans are the same size and then just use the same `i` value in the loop below. This feels contrived.

File: pkg/sql/distsql_plan_backfill.go
Comment (ajwerner) [inline pkg/sql/distsql_plan_backfill.go:120]: this all feels gross. I really don't like `PartitionSpansWithUserData`. Probably with as much code, and fewer allocations than you've infected this code with, you could use a data structure to look up the index:

```go
type spanAndIndex struct {
    roachpb.Span
    idx int
}

var _ interval.Interface = (*spanAndIndex)(nil)

func (si *spanAndIndex) Range() inteval.Range { return si.AsRange() }
func (si *spanAndIndex) ID() uintptr { return uintptr(unsafe.Pointer(si)) }

func (dsp *DistSQLPlanner) createIndexBackfillerMergePhysicalPlan(
	planCtx *PlanningCtx, spec execinfrapb.IndexBackfillMergerSpec, spans [][]roachpb.Span,
) (*PhysicalPlan, error) {
        var n int
	for _, sp := range spans {
		for range sp {
		        n++
		}
	}
	indexSpans := make([]roachpb.Span, 0, n)
        spanIdxs := make([]spanAndIdxs, 0, n)
        spanIdxTree := interval.NewTree(interval.ExclusiveOverlapper)
        for i := range spans {
		for j := range spans[i] {
			indexSpans = append(indexSpans, spans[i][j])
			spanIdxs = append(spanIdxs, spanAndIndex{Span: spans[i][j], idx: i})
			spanIdxTree.Insert(&spanIdxs[len(spanIdxs)-1], true /* fast */)
		}
	}
	spanIdxTree.AdjustRanges()
	spanPartitions, err := dsp.PartitionSpansWithUserData(planCtx, indexSpans)
	// ...
	// now you can look up an index for a span like:
	getIndex := func(sp roachpb.Span) (idx int) {
	    if !spanIdxTree.DoMatching(func(i Interface) (done bool) {
	        i.(*spanAndIndex).idx
	        return false
	     }, sp.AsRange()) {
	        panic("didn't find it")
	     }
	     return idx
	}
	
```

**Review by ajwerner — APPROVED** [review]
Sorry it took me so long to actually make my way through this. My comments are largely superficial and can be made later as cleanup. Let's get this merged so we can gain experience with it

File: pkg/sql/backfill/mvcc_index_merger.go
Comment (stevendanna) [reply pkg/sql/backfill/mvcc_index_merger.go:209]: Happy to update to this now.  Was 16Kb just a bet guess or do you have some thinking there I can include in a comment?

File: pkg/sql/distsql_physical_planner.go
Comment (stevendanna) [reply pkg/sql/distsql_physical_planner.go]: Your other suggestion removes this code completely, so nothing more to do here.

File: pkg/sql/distsql_plan_backfill.go
Comment (stevendanna) [reply pkg/sql/distsql_plan_backfill.go]: Thanks for this suggestion.  I've updated the PR to use this method. @rhu713 Mind taking a quick look at the last few commits to make sure you are OK with those changes.

**Comment by stevendanna** [comment]
bors r=ajwerner

File: pkg/sql/backfill/mvcc_index_merger.go
Comment (stevendanna) [reply pkg/sql/backfill/mvcc_index_merger.go:209]: Decided to push this to a follow up PR. Will open it shortly after this merges.

**Pull request merged**
